python3Packages.beancount: 2.3.6 -> 3.1.0

alapshin opened the pull request:
Add package for beancount v3
Add package for `beangulp` as a replacement for `beancount.ingest`
Update package for `beanquery` to use snapshot version that using TatSu-LTS dependency: https://github.com/beancount/beanquery/issues/226
Update `fava` to version 1.30.1 which supports both `beancount2` and `beancount3`.
## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alapshin committed:
python313Packages.petl: init at 1.7.15

alapshin committed:
python313Package.beangulp: init at 0.2.0

alapshin committed:
fava: 1.29 -> 1.30.1

SharzyL commented:
I am not using v3 currently. How is it working with existing tools now?

alapshin commented:
> I am not using v3 currently. How is it working with existing tools now?

I used v3 last 6 months, and it works ok for me. However, I don't use any custom importers based on `beancount.ingest` which is the main breaking change, and input my expenses by hand. Until recently, Fava didn't support v3, but now it does: https://github.com/beancount/fava/issues/1831

Overall, I think it is in good enough shape for everyday use, but it depends on your setup.

SharzyL commented:
It would be nice to have v3 in nixpkgs. But since there are major incompatibilities between v2 and v3, and v2 is still widely used, I would like to keep both v2 and v3 in nixpkgs, i.e. `beancount` and `beancount3`. Otherwise users may find their tools broken after an upgrade. When v2 is marked as deprecated, we can add warnings in nixpkgs and later replace `beancount` with `beancount3`.

SharzyL commented on pkgs/by-name/fa/fava/package.nix:
```diff
@@ -39,9 +44,19 @@ python3Packages.buildPythonApplication rec {
+    ''
+    # Disable some tests when building with beancount2
+    + (
+      if !isBeancount3 then
```
`lib.optionalString` can simplify the code

alapshin commented on pkgs/by-name/fa/fava/package.nix:
```diff
@@ -1,17 +1,20 @@
   lib,
-  python3Packages,
   fetchPypi,
+  python3Packages,
```
Would it help if, instead of accepting python3Packages, the package accepted beancount dependency directly?

It is possible to build Fava either with v2 or v3, but as far as I understand depending on version used Fava uses different approach for importers under the hood

https://github.com/beancount/fava/blob/main/CHANGES
```
For importers, this sticks to beancount.ingest on v2 and used beangulp for v3.
Due to changes related to duplicate detection, duplicate detection is not
automatically done by Fava here but can still be manually specified as hooks.
``` 

Also, I'm not sure if it is a good idea to mix-match beancount version between different packages (fava, beanprice, beanquery, beangulp, etc.).

SuperSandro2000 replied:
> It is very cumbersome to override python3Packages to build fava with beancount3 due to the strange design in python3Packages.

please see https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#overriding-python-packages-overriding-python-packages

lucasew commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,66 @@
+
+  pythonImportsCheck = [ "beancount" ];
+
+  meta = with lib; {
```
please remove top level with lib

lucasew commented:
Fava works in my Beancount 2.x based database

I didn't migrate it yet.

alapshin committed:
python313Packages.beanquery: 0.1.0 -> 0.1.0-unstable-2025-01-10

alapshin commented:
Had to update beanquery package to a snapshot version which uses TatSu-LTS dependency to fix build, see: https://github.com/beancount/beanquery/issues/226

dotlambda commented:
Beancount was updated to v3 in master, please drop the commit adding beancount3 (which doesn't exist on PyPI under that name anyway).

seberm commented:
Hello @alapshin could you please rebase and drop the commit adding beancount3 (it's already in master). It would be great if we could continue merging this PR. Thanks!

SuperSandro2000 commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,42 @@
+      tables of data.
+    '';
+    license = lib.licenses.mit;
+    mainProgram = "petl";
```
Missing maintainer

SuperSandro2000 commented on pkgs/development/python-modules/beanquery/default.nix:
```diff
@@ -9,16 +9,27 @@
+    src = fetchFromGitHub {
+      owner = "dnicolodi";
+      repo = "TatSu-LTS";
+      version = "5.13.1";
+      tag = "v5.13.1-LTS";
```
```suggestion
      tag = "v5.13.1-LTS";
```

SuperSandro2000 commented on pkgs/development/python-modules/beanquery/default.nix:
```diff
@@ -9,16 +9,27 @@
   tatsu,
 }:
+let
+  tatsu-lts = tatsu.overrideAttrs (oldAttrs: {
```
```suggestion
  tatsu-lts = tatsu.overrideAttrs (oldAttrs: {
    version = "5.13.1";
```

SuperSandro2000 commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,66 @@
+    hash = "sha256-z2aGhpx+o+78CU7hPthmv196K7DGHk1PXfPjX4Rs/98=";
+  };
+
+  # Tests require files not included in the PyPI archive.
+  # Also there is an import error after migration to meson build.
```
Then do not download from pypi. This is a very common issue.

alapshin replied:
First line is copied `as-is` from existing package for beancount v2, not sure if it still applies to v3. Second one is added when trying to package v3.

There are a number of failed tests with the same error message
```
______________ ERROR collecting beancount/core/account_test.py ________________
ImportError while importing test module '/build/beancount-3.0.0/beancount/core/account_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/0l539chjmcq5kdd43j6dgdjky4sjl7hl-python3-3.12.8/lib/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
beancount/__init__.py:22: in <module>
    from .api import *  # pylint: disable=wildcard-import
beancount/api.py:26: in <module>
    from .loader import load_file, load_encrypted_file, load_doc
beancount/loader.py:28: in <module>
    from beancount.parser import parser
beancount/parser/parser.py:120: in <module>
    from beancount.parser import _parser
E   ImportError: cannot import name '_parser' from 'beancount.parser' (/build/beancount-3.0.0/beancount/parser/__init__.py)
```

I'm don't have much experience with Python and/or Meson build system and not sure how to address it. Considering that v2 already had tests disabled, I disabled them for v3 as well. I understand that it would be better to fix test failure, but not sure how.

SuperSandro2000 commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,66 @@
+  pname = "beancount";
+  pyproject = true;
+
+  disabled = !isPy3k;
+
```
```suggestion
```
we don't support python 2 at all anymore

alapshin commented:
Update Beancount to 3.1.0

dotlambda commented:
> please drop the commit adding beancount3

Still there.

SharzyL commented:
> > please drop the commit adding beancount3
> 
> Still there.

No need to drop since #382819 is merged

dotlambda commented:
`beancount3` does not exist on PyPI so it can't be in `python3Packages`

ambroisie commented:
From the [contributing guidelines for python packaging](https://nixos.org/manual/nixpkgs/stable/#contributing-guidelines):

> Packages from sources such as GitHub and GitLab that do not exist on PyPI should not use a name that is already used on PyPI. When possible, they should use the package repository name prefixed with the owner (e.g. organization) name and using a - as delimiter.

My reading is that it is not _necessary_ for a package to be on PyPI to be included in `python3Packages`,

dotlambda commented:
If absolutely necessary, beancount_3 would be acceptable. But I would prefer to update beancount.

dotlambda commented on pkgs/by-name/fa/fava/package.nix:
```diff
@@ -39,17 +41,22 @@ python3Packages.buildPythonApplication rec {
+    ''
+    # Disable some tests when building with beancount2
+    + lib.optionalString (lib.versions.major python3Packages.beancount.version == "2") ''
+      export SNAPSHOT_IGNORE=true
```
Let's use `env`.

dotlambda commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,70 @@
+  pname = "beancount";
+  pyproject = true;
+
+  src = fetchPypi {
```
`fetchFromGitHub` is preferred

dotlambda commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,70 @@
+      generate a variety of reports from them, and provides a web interface.
+    '';
+    license = lib.licenses.gpl2Only;
+    maintainers = with lib.maintainers; [ bhipple ];
```
This should be yourself.

dotlambda commented on pkgs/development/python-modules/beancount/default.nix:
Separate commit for changes to this file.

dotlambda commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,45 @@
+    "petl"
+  ];
+
+  meta = {
```
changelog missing

dotlambda commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,45 @@
+  version = "1.7.15";
+  pyproject = true;
+
+  src = fetchPypi {
```
`fetchFromGitHub` is preferred

dotlambda commented on pkgs/development/python-modules/petl/default.nix:
Tests missing.

alapshin replied:
Which ones?

dotlambda replied:
e.g. `pytestCheckHook`

alapshin replied:
It does have `pytestCheckHook`

dotlambda replied:
Oops, I overlooked that.

alapshin commented on pkgs/development/python-modules/beangulp/default.nix:
```diff
@@ -0,0 +1,57 @@
+    "beangulp"
+  ];
+
+  meta = {
```
Where it should point? Repo doesn't have separate changelog file and because it is snapshot version there is no tag release to point to either.

dotlambda replied:
I see. In that case there's no need for a changelog.

alapshin commented on pkgs/development/python-modules/beanquery/default.nix:
```diff
@@ -9,16 +9,27 @@
   tatsu,
 }:
+let
+  tatsu-lts = tatsu.overrideAttrs (oldAttrs: {
```
Do I need to create separate package for tatsu-lts?

dotlambda replied:
No, you cannot use a different version of tatsu.

alapshin replied:
> No, you cannot use a different version of tatsu.

Even if it is fork of original package: https://github.com/dnicolodi/TatSu-LTS ?

Beanquery requires it according to https://github.com/beancount/beanquery/issues/226

alapshin replied:
Added a new package as `tatsu-lts` but now `beanquery` fails to build with

```
       > Running phase: pythonRuntimeDepsCheckHook
       > Executing pythonRuntimeDepsCheck
       > Checking runtime dependencies for beanquery-0.2.0.dev0-py3-none-any.whl
       >   - tatsu-lts not installed
       For full logs, run 'nix log /nix/store/f3b2b59lk9hw9d75j0nyaqp1lqmgdi8c-python3.12-beanquery-0.2.0-unstable-2025-01-10.drv'.
```

Not sure why and will have to investigate.

alapshin replied:
Had to use to correct git tag to point to LTS release

dotlambda commented on pkgs/development/python-modules/beangulp/default.nix:
```diff
@@ -0,0 +1,58 @@
+}:
+buildPythonPackage rec {
+  pname = "beangulp";
+  version = "0.2.0-unstable-2025-01-02";
```
What's the issue with 0.2.0?

alapshin replied:
None, didn't noticed that new release was tagged, updated package to use it.

dotlambda commented:
How about we put fava and beangulp in separate PRs? Those could be merged before we figure out how to deal with v2 vs v3.

alapshin commented:
> How about we put fava and beangulp in separate PRs? Those could be merged before we figure out how to deal with v2 vs v3.

I can move beancount3 commit to separate PR. But other commits have to be in one PR, because new `fava` version depends on `beanquery` and `beangulp` and `beangulp` requires `petl` dependency.

dotlambda commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,48 @@
+
+  meta = {
+    homepage = "https://github.com/petl-developers/petl";
+    changelog = "https://github.com/petl-developers/petl/releases/tag/${src.tag}/CHANGELOG.md";
```
doesn't exist

alapshin committed:
python313Packages.tatsu-lts: init at 5.13.1

dotlambda commented on pkgs/development/python-modules/tatsu-lts/default.nix:
```diff
@@ -0,0 +1,46 @@
+    hash = "sha256-iZtYqPvQxXl6SFG2An7dN3KxaxCTvAiAkeeuXUhLuF0=";
+  };
+
+  nativeBuildInputs = [ setuptools ];
```
build-system

dotlambda commented on pkgs/development/python-modules/beanquery/default.nix:
```diff
@@ -7,18 +7,18 @@
+buildPythonPackage {
   pname = "beanquery";
-  version = "0.1.0";
+  version = "0.2.0-unstable-2025-01-10";
```
Doesn't follow guidelines

alapshin replied:
I believe it does.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#versioning

```
If a package is a commit from a repository without a version assigned, then the version attribute should be the latest upstream version preceding that commit, followed by -unstable- and the date of the (fetched) commit. The date must be in "YYYY-MM-DD" format.
```

dotlambda replied:
Is there a 0.2.0?

alapshin replied:
Changed to 0.1.0-unstable-2025-01-10

alapshin commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,48 @@
+      tables of data.
+    '';
+    license = lib.licenses.mit;
+    mainProgram = "petl";
```
Yes, https://github.com/petl-developers/petl/blob/1e77626f17f6a89697abd17fd9d98b2ad9ed41cb/setup.py#L13

dotlambda commented on pkgs/development/python-modules/beangulp/default.nix:
```diff
@@ -0,0 +1,58 @@
+  build-system = [ setuptools ];
+
+  dependencies = [
+    beancount
+    chardet
```
```suggestion
    beancount
    beautifulsoup4
    chardet
```

dotlambda commented on pkgs/development/python-modules/beangulp/default.nix:
```diff
@@ -0,0 +1,58 @@
+  nativeCheckInputs = [
+    petl
+    pytestCheckHook
+    regex
```
Are you sure regex is necessary?

dotlambda commented on pkgs/by-name/fa/fava/package.nix:
```diff
@@ -39,17 +41,23 @@ python3Packages.buildPythonApplication rec {
   preCheck = ''
-    export HOME=$TEMPDIR
+    # Needed for tests
+    HOME = "$(mktemp -d)";
```
That's not gonna work.

alapshin commented on pkgs/development/python-modules/beancount/3.nix:
```diff
@@ -0,0 +1,70 @@
+    pytestCheckHook
+  ];
+
+  propagatedBuildInputs = [
```
Removed `propagatedBuildInputs`

dotlambda commented on pkgs/by-name/fa/fava/package.nix:
```diff
@@ -24,6 +24,8 @@ python3Packages.buildPythonApplication rec {
   dependencies = with python3Packages; [
```
Remove jaraco-functools from these.

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15960,6 +15960,7 @@ with pkgs;
   samplv1 = libsForQt5.callPackage ../applications/audio/samplv1 { };
 
   beancount = with python3.pkgs; toPythonApplication beancount;
+  beancount3 = with python3.pkgs; toPythonApplication beancount3;
```
```suggestion

  beancount_3 = with python3.pkgs; toPythonApplication beancount_3;
```

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1587,6 +1587,8 @@ self: super: with self; {
 
   beancount = callPackage ../development/python-modules/beancount { };
 
+  beancount3 = callPackage ../development/python-modules/beancount/3.nix { };
```
```suggestion
  beancount_3 = callPackage ../development/python-modules/beancount/3.nix { };
```

dotlambda commented on pkgs/development/python-modules/petl/default.nix:
```diff
@@ -0,0 +1,49 @@
+  meta = {
+    homepage = "https://github.com/petl-developers/petl";
+    changelog = "https://github.com/petl-developers/petl/releases/tag/${src.tag}";
+    description = "Python Extract Transform and Load Tables of Data";
+    longDescription = ''
+      A general purpose Python package for extracting, transforming and loading
+      tables of data.
+    '';
```
```suggestion
    description = "Python package for extracting, transforming and loading tables of data";
```

dotlambda commented on pkgs/development/python-modules/beanquery/default.nix:
```diff
@@ -7,18 +7,18 @@
+buildPythonPackage {
   pname = "beanquery";
-  version = "0.1.0";
+  version = "0.1.0-unstable-2025-01-10";
```
Can we ask upstream to cut a release?

alapshin replied:
> Can we ask upstream to cut a release?

Someone already did at https://github.com/beancount/beanquery/issues/226#issuecomment-2614649306

Upstream suggested to use version 0.1.0 that requires `tatsu >= 5.7.4, < 5.8.0` which is not present at nixpkgs, using `pythonRelaxDepsHook` doesn't work, because beanquery fails to build with original tatsu v5.13.1 (non-LTS one).

dotlambda replied:
You can use `substituteInPlace` to replace tatsu with tatsu-lts.

alapshin replied:
> You can use `substituteInPlace` to replace tatsu with tatsu-lts.

Wouldn't help because there are API changes in `tatsu` v5.13.0 and later that are present in lts version as well. These change cause all `beanquery` tests to fail with `AttributeError`. The only way to package `beanquery` at `v0.1.0` is to pin `tatsu` dependency at previous version `v5.12.2`

alapshin commented:
Removed `pytestCheckHook` from beancount v3 package because using

```
preCheck = ''
  # import from $out
  rm -r beancount
''  
```
left no tests to run and caused pytest to fail.

alapshin commented:
One more argument for dropping support for beancount v2

New version of beanprice doesn't support it either: https://github.com/beancount/beanprice/pull/98

dotlambda commented:
> Removed `pytestCheckHook` from beancount v3 package because using
> 
> ```
> preCheck = ''
>   # import from $out
>   rm -r beancount
> ''  
> ```
> 
> left no tests to run and caused pytest to fail.

Then do something different like renaming the directory or copy the tests before removing it.

dotlambda commented:
> Separate commit for changes to this file.

This still hasn't been addressed.
The attribute still has the wrong nsme.
...

alapshin commented:
> One more argument for dropping support for beancount v2
> 
> New version (2.0.0) of beanprice doesn't support it either: [beancount/beanprice#98](https://github.com/beancount/beanprice/pull/98)

After some thought I really want to drop support for beancount v2 in this PR but not sure which way forward is the best.

I understand that there are people who would be inconvenienced by this but feel now is a good moment to od it. Most of the tools from the ecosystem supports beancount v3 and some supports only v3 going forward (see beanprice for example).

I see multiple ways forward

1. Have only v3
2. Have both beancount v2 and v3 in nixpkgs and use v2 by default
3. Have both beancount v2 and v3 in nixpkgs and use v3 by default

In both situation 2 and 3 it should be possible to force version using overlay similar to
```
    pythonPackagesExtensions = prev.pythonPackagesExtensions ++ [
      (pyfinal: pyprev: {
        beancount = pyprev.beancount_3; case 2
        // or
        beancount = pyprev.beancount_2; // case 3
      })
    ]
```
and have all other packages use this version. However this only works as long as package itself implements support for v2 and v3 (like fava for example). If package drops support for v2 (like beanprice 2.0) it wouldn't work and one would have to have separate beanprice package for beancount v2. This could be achievable but I'm not sure if it worth it.

alapshin commented:
> > Separate commit for changes to this file.
> 
> This still hasn't been addressed. The attribute still has the wrong name. ...

I have removed formatting changes to `beancount/default.nix`. Attribute was also renamed from `beancount3` to `beancount_3`.

dotlambda commented:
> Have both beancount v2 and v3 in nixpkgs and use v3 by default

Let's do this one for now. We can drop v2 once it fails to build.

alapshin committed:
beanprice: 1.2.1-unstable-2024-06-19 -> 2.0.0

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1587,7 +1587,9 @@ self: super: with self; {
   before-after = callPackage ../development/python-modules/before-after { };
 
-  beancount = callPackage ../development/python-modules/beancount { };
+  beancount = callPackage ../development/python-modules/beancount/3.nix { };
+
+  beancount_2 = callPackage ../development/python-modules/beancount { };
```
```suggestion
  beancount = callPackage ../development/python-modules/beancount { };

  beancount_2 = callPackage ../development/python-modules/beancount/2.nix { };
```

dotlambda reviewed (commented):
The commit message needs to be adapted. I suggest

* `python313Packages.beancount: 2.3.6 -> 3.1.0`
* `python313Packages.beancount_2: init at 2.3.6`

alapshin commented:
> > Have both beancount v2 and v3 in nixpkgs and use v3 by default
> 
> Let's do this one for now. We can drop v2 once it fails to build.

Done. Now `beancount` points to a v3 package. 

Not sure if I need to rename
```
beancount/default.nix -> beancount/2.nix
beancount/3.nix -> beancount/default.nix
```

Also updated beanprice 2.0 which require beancount >= 3, but because v3 is default now it works.

dotlambda commented:
> Also updated beanprice 2.0 which require beancount >= 3, but because v3 is default now it works.

It might make sense to set `meta.broken = lib.versionOlder beancount.version "3"`

alapshin committed:
python313Packages.beancount_2: init at 2.3.6

alapshin committed:
python3Packages.beancount: 2.3.6 -> 3.1.0

dotlambda commented on pkgs/by-name/be/beanprice/package.nix:
```diff
@@ -1,19 +1,20 @@
 {
   lib,
+  beancount,
```
```suggestion
```

dotlambda commented on pkgs/by-name/be/beanprice/package.nix:
```diff
@@ -34,6 +35,7 @@ python3Packages.buildPythonApplication rec {
   pythonImportsCheck = [ "beancount" ];
 
   meta = {
+    broken = lib.versionOlder beancount.version "3";
```
```suggestion
    broken = lib.versionOlder python3Packages.beancount.version "3";
```

dotlambda commented on pkgs/by-name/be/beanprice/package.nix:
```diff
@@ -34,6 +35,7 @@ python3Packages.buildPythonApplication rec {
   pythonImportsCheck = [ "beancount" ];
```
Should this be beanprice?

dotlambda commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1587,7 +1587,9 @@ self: super: with self; {
   before-after = callPackage ../development/python-modules/before-after { };
 
-  beancount = callPackage ../development/python-modules/beancount { };
+  beancount = callPackage ../development/python-modules/beancount/3.nix { };
```
```suggestion
  beancount = callPackage ../development/python-modules/beancount { };
```

alapshin commented:
Added comment in `beanquery` package explaining choice of snapshot version.

dotlambda commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 371541`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 37 packages built:</summary>
  <ul>
    <li>beancount (python312Packages.beancount)</li>
    <li>beancount.dist (python312Packages.beancount.dist)</li>
    <li>beancount_2 (python312Packages.beancount_2)</li>
    <li>beancount_2.dist (python312Packages.beancount_2.dist)</li>
    <li>beanprice</li>
    <li>beanprice.dist</li>
    <li>beanquery (python312Packages.beanquery)</li>
    <li>beanquery.dist (python312Packages.beanquery.dist)</li>
    <li>fava</li>
    <li>fava.dist</li>
    <li>ledger2beancount</li>
    <li>python312Packages.beancount-docverif</li>
    <li>python312Packages.beancount-docverif.dist</li>
    <li>python312Packages.beancount-plugin-utils</li>
    <li>python312Packages.beancount-plugin-utils.dist</li>
    <li>python312Packages.beangulp</li>
    <li>python312Packages.beangulp.dist</li>
    <li>python312Packages.petl</li>
    <li>python312Packages.petl.dist</li>
    <li>python312Packages.tatsu-lts</li>
    <li>python312Packages.tatsu-lts.dist</li>
    <li>python313Packages.beancount</li>
    <li>python313Packages.beancount-docverif</li>
    <li>python313Packages.beancount-docverif.dist</li>
    <li>python313Packages.beancount-plugin-utils</li>
    <li>python313Packages.beancount-plugin-utils.dist</li>
    <li>python313Packages.beancount.dist</li>
    <li>python313Packages.beancount_2</li>
    <li>python313Packages.beancount_2.dist</li>
    <li>python313Packages.beangulp</li>
    <li>python313Packages.beangulp.dist</li>
    <li>python313Packages.beanquery</li>
    <li>python313Packages.beanquery.dist</li>
    <li>python313Packages.petl</li>
    <li>python313Packages.petl.dist</li>
    <li>python313Packages.tatsu-lts</li>
    <li>python313Packages.tatsu-lts.dist</li>
  </ul>
</details>

alapshin commented:
@dotlambda Thanks for your patience and feedback while moving this PR forward.
